nifi-nar-bundles/nifi-stateless-processor-bundle/nifi-stateless-processor/src/main/java/org/apache/nifi/processors/stateless/ExecuteStateless.java

```diff
@@ -0,0 +1,1202 @@
+            results.add(rpg.getTargetUris(), "RPG Target URI");
+            results.add(rpg.getComments(), "RPG Comments");
+            results.add(rpg.getIdentifier(), "RPG Identifier");
+
+            for (final VersionedRemoteGroupPort port : rpg.getInputPorts()) {
+                results.add(port.getName(), "RPG Input Port Name");
+                results.add(port.getIdentifier(), "RPG Input Port ID");
+                results.add(port.getTargetId(), "RPG Input Port Target ID");
+            }
+
+            for (final VersionedRemoteGroupPort port : rpg.getOutputPorts()) {
+                results.add(port.getName(), "RPG Output Port Name");
+                results.add(port.getIdentifier(), "RPG Output Port ID");
+                results.add(port.getTargetId(), "RPG Output Port Target ID");
+            }
+        }
+        for (final Map.Entry<String, String> entry : group.getVariables().entrySet()) {
+            results.add(entry.getKey(), "Variable Name");
+            results.add(entry.getValue(), "Value of Variable " + entry.getKey());
+        }
+        results.add(group.getParameterContextName(), "Parameter Context Name");
+
+        for (final VersionedConnection connection : group.getConnections()) {
+            results.add(connection.getIdentifier(), "Connection ID");
+            results.add(connection.getName(), "Connection Name");
+            if (connection.getSelectedRelationships() != null) {
+                results.add(connection.getSelectedRelationships().toString(), "Selected Relationships");
+            }
+            results.add(connection.getComments(), "Connection Comments");
+        }
+        for (final VersionedControllerService service : group.getControllerServices()) {
+            results.add(service.getName(), "Controller Service Name");
+            results.add(service.getType(), "Controller Service Type");
+            results.add(service.getIdentifier(), "Controller Service ID");
+
+            for (final Map.Entry<String, String> entry : service.getProperties().entrySet()) {
+                results.add(entry.getKey(), "Controller Service Property Name");
+                results.add(entry.getValue(), "Value of Controller Service Property " + entry.getKey());
+            }
+            results.add(service.getComments(), "Controller Service Comments");
+
+            final Bundle bundle = service.getBundle();
+            if (bundle != null) {
+                results.add(bundle.getGroup(), "Bundle Group ID for Controller Service " + service.getType());
+                results.add(bundle.getArtifact(), "Bundle Artifact ID for Controller Service " + service.getType());
+                results.add(bundle.getVersion(), "Bundle Version for Controller Service " + service.getType());
+            }
+        }
+        for (final VersionedProcessGroup child : group.getProcessGroups()) {
+            search(child, results);
+        }
+    }
+
+
+    private interface DataflowRetrieval {
```

Comment:
This internal interface and associated implementations keep everything together, but they also make it more difficult to unit test. What do you think about breaking out the interface and retrieval implementations to a separate package that could be unit-tested separately?

Reply:
Fair enough.

Reply:
On further consideration, what do you think about turning this into a Controller Service interface?  File-based and Registry-based implementations should cover most use cases, but that would allow for custom implementations using other approaches.  That would also remove the direct dependency on nifi-registry, and also remove the Registry SSL Context Service property.

Reply:
I actually had the same thought when initially reading through the code.  Might such a controller service be useful outside of ExecuteStateless?  Not sure if this would be overkill, but worth considering.

Reply:
I would avoid controller services in this situation. The advantage to using a controller service would be that you could configure it once and reuse it many times for many instances of ExecuteStateless. Except that for registry you'd still need a different bucket, flow, version every time. All you'd really be able to reuse is the timeout, url, and maybe SSL Context. So the benefits of a controller service here, IMO, are outweighed but the complexity of having to configure that service.

Reply:
Could the the `Dataflow File/URL` property be repurposed to parameterize the Bucket, Flow, and Version properties? I understand the tradeoff of complexity, but the current implementation requires a number of properties that are specific to Registry.  The `dependsOn` element helps keep this more manageable, but it does add a good bit of code complexity.  It seems like another potential implementation could be direct retrieval from a Git repository or web service. If those additional implementations seem unlikely, or would not introduce too much complication to the current approach, then a Controller Service seems unnecessary.